From 03056717e85548530030175de5feed626be1602a Mon Sep 17 00:00:00 2001
From: Anthony Mallet <anthony.mallet@laas.fr>
Date: Tue, 8 Jun 2021 14:33:31 +0200
Subject: [PATCH] Prevent posterRead() from reading uninitialized data

posterTake() in POSTER_READ mode now ensures that data is initialized by
checking the H2DEV_POSTER_FLG_FRESH flag.

It will return the EMPTY_POSTER error if the data is not initialized.

Closes #318
---
 src/posterLib/localPosterLib.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git src/posterLib/localPosterLib.c src/posterLib/localPosterLib.c
index 087227c..7cd770c 100644
--- src/posterLib/localPosterLib.c
+++ src/posterLib/localPosterLib.c
@@ -325,10 +325,25 @@ localPosterTake(POSTER_ID posterId, POSTER_OP op)
 	errnoSet(S_posterLib_POSTER_CLOSED);
 	return(ERROR);
     }
-    if (op == POSTER_WRITE && H2DEV_POSTER_TASK_ID(dev) != getpid()) {
-	errnoSet(S_posterLib_NOT_OWNER);
-	return ERROR;
+
+    switch (op) {
+	case POSTER_WRITE:
+	    if (H2DEV_POSTER_TASK_ID(dev) != getpid()) {
+		errnoSet(S_posterLib_NOT_OWNER);
+		return ERROR;
+	    }
+	    break;
+
+	case POSTER_READ:
+	    if (H2DEV_POSTER_FLG_FRESH(dev) != TRUE) {
+		errnoSet(S_posterLib_EMPTY_POSTER);
+		return ERROR;
+	    }
+	    break;
+
+	default: break;
     }
+
     if (h2semTake(H2DEV_POSTER_SEM_ID(dev), WAIT_FOREVER) == FALSE) {
 	return ERROR;
     }
-- 
2.25.1

